File: BudgetPro.Server/Services/CategoryService.cs
using BudgetPro.Server.Data;
using BudgetPro.Server.DTOs;
using BudgetPro.Server.Interfaces;
using BudgetPro.Server.Requests.AddRequests;
using BudgetPro.Server.Requests.GetRequests;
using Microsoft.EntityFrameworkCore;

namespace BudgetPro.Server.Services;

public class CategoryService : ICategoryService
{
    private readonly IDbContextFactory<BudgetDbContext> dbContextFactory;

    public CategoryService(IDbContextFactory<BudgetDbContext> dbContextFactory)
    {
        this.dbContextFactory = dbContextFactory;
    }

    public async Task<CategoryDTO> AddCategory(AddCategoryRequest request)
    {
        if (request is null)
        {
            return new CategoryDTO();
        }

        using var context = await dbContextFactory.CreateDbContextAsync();

        var possibleCategory = await context.Categories
            .Where(x => x.Name == request.Name)
            .FirstOrDefaultAsync();

        if (possibleCategory is not null)
        {
            return new CategoryDTO();
        }

        Category newCategory = new Category()
        {
            IsIncome = request.IsIncome,
            Name = request.Name ?? "",
            UserId = request.UserId,
        };

        await context.Categories.AddAsync(newCategory);
        await context.SaveChangesAsync();

        return newCategory.ToDTO();
    }

    public async Task<List<CategoryDTO>> GetAllCategories()
    {
        using var context = await dbContextFactory.CreateDbContextAsync();

        return await context.Categories
            .Select(x => x.ToDTO())
            .ToListAsync();
    }

    public async Task<CategoryDTO> GetCategoryById(FindCategoryRequest request)
    {
        using var context = await dbContextFactory.CreateDbContextAsync();

        return await context.Categories
            .Where(x => x.Id == request.Id)
            .Select(x => x.ToDTO())
            .FirstOrDefaultAsync() ?? new CategoryDTO();
    }

    public async Task<CategoryDTO> GetCategoryByName(FindCategoryRequest request)
    {
        using var context = await dbContextFactory.CreateDbContextAsync();

        return await context.Categories
            .Where(x => x.Name == request.Name && x.UserId == request.UserId)
            .Select(x => x.ToDTO())
            .FirstOrDefaultAsync() ?? new CategoryDTO();
    }


}

File: budgetpro.client/src/app/services/category/category.service.ts
import { Injectable } from '@angular/core';
import { environment } from '../../../environments/environment';
import { HttpClient } from '@angular/common/http';

export interface Category {
  id: number;
  name: string;
  isIncome: boolean;
  userId: number;
}

export interface FindCategoryRequest {
  id: number;
  name: string;
  userId: number;
}

export interface AddCategoryRequest {
  name: string;
  isIncome: boolean;
  userId: number;
}

@Injectable({
  providedIn: 'root',
})
export class CategoryService {
  private readonly apiUrl = environment.apiUrl;

  constructor(private httpClient: HttpClient) {}

  getAllCategories() {
    return this.httpClient.get<Category[]>(`${this.apiUrl}/Category/GetAll`);
  }

  getCategoryById(request: FindCategoryRequest) {
    return this.httpClient.post<Category>(
      `${this.apiUrl}/Category/GetById}`,
      request
    );
  }

  addNewCategory(request: AddCategoryRequest) {
    return this.httpClient.post<Category>(
      `${this.apiUrl}/Category/Add`,
      request
    );
  }
}
